spring-rabbit/src/main/java/org/springframework/amqp/rabbit/listener/adapter/ReplyingMessageListener.java

```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright 2016 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.amqp.rabbit.listener.adapter;
+
+/**
+ * A Message Listener that returns a reply - intended for lambda use in a
+ * {@link MessageListenerAdapter}.
+ *
+ * @author Gary Russell
+ * @since 2.0
+ *
+ */
+@FunctionalInterface
+public interface ReplyingMessageListener<T, R> {
+
+	/**
+	 * Handle the message and return a reply.
+	 * @param t the request.
+	 * @return the reply.
+	 */
+	R handleMessage(T t);
```

Comment:
Good point, generally - although, since the MLA takes an `Object` ctor arg, we always have to cast the lambda so I thought this 

    new MessageListenerAdapter((ReplyingMessageListener<String, String>) in -> in.toUpperCase()));

would add some semantic sugar over

    new MessageListenerAdapter((Function<String, String>) in -> in.toUpperCase()));

Also, the MLA uses the method name `handleMessage` by default.

I suppose we could add code to call `apply` if it's a `Function` but then what if someone has an existing listener with a `handleMessage` method that also implements `Function` ? We'd start calling the "wrong" method.